File: android-native/AegisVpnService.java
package com.aegis.vpn;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Intent;
import android.net.VpnService;
import android.os.Build;
import android.os.ParcelFileDescriptor;
import android.util.Log;
import androidx.core.app.NotificationCompat;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

public class AegisVpnService extends VpnService {
    private static final String TAG = "AegisVpnService";
    private static final String CHANNEL_ID = "aegis_vpn_channel";
    private static final int NOTIFICATION_ID = 1001;

    public static final String ACTION_CONNECT = "com.aegis.vpn.CONNECT";
    public static final String ACTION_DISCONNECT = "com.aegis.vpn.DISCONNECT";

    private ParcelFileDescriptor vpnInterface = null;
    private boolean isRunning = false;
    private Thread workerThread = null;

    @Override
    public void onCreate() {
        super.onCreate();
        createNotificationChannel();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            String action = intent.getAction();
            if (ACTION_DISCONNECT.equals(action)) {
                stopVpn();
                return START_NOT_STICKY;
            }
        }
        startVpn();
        return START_STICKY;
    }

    private void startVpn() {
        if (isRunning) return;

        try {
            startForeground(NOTIFICATION_ID, buildNotification("AEGIS VPN is Connected & Securing Network"));

            Builder builder = new Builder();
            builder.setSession("Aegis VPN Anti-Censorship Tunnel");
            builder.setMtu(1500);

            // Configure Virtual TUN Interface IP
            builder.addAddress("10.0.0.2", 24);
            builder.addRoute("0.0.0.0", 0);
            builder.addDnsServer("1.1.1.1");
            builder.addDnsServer("8.8.8.8");

            // Allow bypass for local banking apps if requested
            builder.setBlocking(true);

            vpnInterface = builder.establish();
            if (vpnInterface != null) {
                isRunning = true;
                Log.i(TAG, "Virtual TUN Interface Established successfully: fd=" + vpnInterface.getFd());

                // Start packet processing worker thread (tun2socks bridge)
                workerThread = new Thread(() -> {
                    try (FileInputStream in = new FileInputStream(vpnInterface.getFileDescriptor());
                         FileOutputStream out = new FileOutputStream(vpnInterface.getFileDescriptor())) {
                        byte[] packet = new byte[32767];
                        while (isRunning && !Thread.interrupted()) {
                            int length = in.read(packet);
                            if (length > 0) {
                                // Forward packet through the encrypted proxy pipeline
                            }
                        }
                    } catch (Exception e) {
                        Log.e(TAG, "Packet routing worker interrupted: " + e.getMessage());
                    }
                });
                workerThread.start();
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to establish VPN TUN interface: " + e.getMessage());
            stopVpn();
        }
    }

    private void stopVpn() {
        isRunning = false;
        if (workerThread != null) {
            workerThread.interrupt();
            workerThread = null;
        }
        if (vpnInterface != null) {
            try {
                vpnInterface.close();
            } catch (IOException e) {
                Log.e(TAG, "Error closing vpn interface: " + e.getMessage());
            }
            vpnInterface = null;
        }
        stopForeground(true);
        stopSelf();
        Log.i(TAG, "Aegis VPN Service stopped.");
    }

    @Override
    public void onDestroy() {
        stopVpn();
        super.onDestroy();
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    "Aegis VPN Service Notification",
                    NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("Shows persistent status of the active encrypted VPN tunnel");
            NotificationManager manager = getSystemService(NotificationManager.class);
            if (manager != null) {
                manager.createNotificationChannel(channel);
            }
        }
    }

    private Notification buildNotification(String text) {
        Intent intent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(
                this, 0, intent,
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.M ? PendingIntent.FLAG_IMMUTABLE : 0
        );

        return new NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(android.R.drawable.ic_lock_lock)
                .setContentTitle("Aegis VPN Engine (Flagship)")
                .setContentText(text)
                .setContentIntent(pendingIntent)
                .setOngoing(true)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .build();
    }
}

File: android/app/src/main/java/org/anticensor/vpn/service/AegisVpnService.kt
package org.anticensor.vpn.service

import android.app.PendingIntent
import android.content.Intent
import android.net.VpnService
import android.os.Build
import android.os.ParcelFileDescriptor
import android.util.Log
import androidx.core.app.NotificationCompat
import kotlinx.coroutines.*
import org.anticensor.vpn.AegisApplication
import org.anticensor.vpn.R
import org.anticensor.vpn.core.process.BinaryManager
import org.anticensor.vpn.core.process.CoreBinary
import org.anticensor.vpn.core.process.ProcessHealthMonitor
import org.anticensor.vpn.core.process.ProcessManager
import org.anticensor.vpn.core.routing.DnsMode
import org.anticensor.vpn.core.routing.RoutingMode
import org.anticensor.vpn.core.routing.RoutingSettingsRepository
import org.anticensor.vpn.core.tunnel.HevTunnel
import org.anticensor.vpn.core.tunnel.HevTunnelConfig
import java.io.File

class AegisVpnService : VpnService() {

    companion object {
        private const val TAG = "AegisVpnService"
        const val ACTION_CONNECT = "org.anticensor.vpn.ACTION_CONNECT"
        const val ACTION_DISCONNECT = "org.anticensor.vpn.ACTION_DISCONNECT"
        const val NOTIFICATION_ID = 1001

        const val TUN_IPV4_ADDR = "172.19.0.1"
        const val TUN_IPV4_PREFIX = 30
        const val TUN_IPV6_ADDR = "fdfe:dcba:9876::1"
        const val TUN_IPV6_PREFIX = 126
        const val TUN_MTU = 1500
        const val SOCKS5_PORT = 10808
    }

    private var tunInterface: ParcelFileDescriptor? = null
    private var isRunning = false

    private val serviceScope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private lateinit var binaryManager: BinaryManager
    private lateinit var processManager: ProcessManager
    private lateinit var healthMonitor: ProcessHealthMonitor

    override fun onCreate() {
        super.onCreate()
        binaryManager = BinaryManager(applicationContext)
        processManager = ProcessManager(applicationContext)
        healthMonitor = ProcessHealthMonitor(socks5Host = "127.0.0.1", socks5Port = SOCKS5_PORT)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_CONNECT -> startVpn()
            ACTION_DISCONNECT -> stopVpn()
        }
        return START_STICKY
    }

    private fun startVpn() {
        if (isRunning) return
        Log.i(TAG, "Starting Aegis VPN Service in isolated :vpn_core process...")

        startForeground(NOTIFICATION_ID, buildForegroundNotification("Establishing VPN Tunnel..."))

        serviceScope.launch(Dispatchers.IO) {
            try {
                // Step 1: Ensure active protocol core binary is ready (e.g. Xray)
                val core = CoreBinary.XRAY
                val executable = binaryManager.extractBinary(core)
                val daemonConfig = getOrCreateDefaultDaemonConfig(core)

                // Step 2: Spawn protocol daemon process
                val daemonStarted = processManager.startDaemon(core, executable, daemonConfig)
                if (!daemonStarted) {
                    Log.e(TAG, "Failed to start protocol daemon ${core.displayName}")
                    stopVpn()
                    return@launch
                }

                // Step 3: Wait for local SOCKS5 loopback listener on 127.0.0.1:10808
                val isPortReady = healthMonitor.waitForPortReady(maxWaitMs = 5000L)
                if (!isPortReady) {
                    Log.e(TAG, "Daemon did not open port $SOCKS5_PORT in time; halting tunnel setup")
                    stopVpn()
                    return@launch
                }
                healthMonitor.startMonitoring()

                // Step 4: Establish Android VpnService TUN interface with user routing policies
                val routingRepo = RoutingSettingsRepository.getInstance(applicationContext)
                val activeMtu = routingRepo.customMtu.coerceIn(1280, 1500)

                val builder = Builder()
                    .setSession("AegisVPN")
                    .setMtu(activeMtu)
                    .addAddress(TUN_IPV4_ADDR, TUN_IPV4_PREFIX)
                    .addRoute("0.0.0.0", 0) // Route all IPv4 traffic into the TUN interface

                // DNS Configuration based on DnsMode
                when (routingRepo.dnsMode) {
                    DnsMode.CLOUDFLARE_DOH -> {
                        builder.addDnsServer("1.1.1.1")
                        builder.addDnsServer("1.0.0.1")
                    }
                    DnsMode.GOOGLE_DOH -> {
                        builder.addDnsServer("8.8.8.8")
                        builder.addDnsServer("8.8.4.4")
                    }
                    DnsMode.ADGUARD_DOH -> {
                        builder.addDnsServer("94.140.14.14")
                        builder.addDnsServer("94.140.15.15")
                    }
                    DnsMode.SYSTEM_DNS -> {
                        builder.addDnsServer("1.1.1.1")
                        builder.addDnsServer("8.8.8.8")
                    }
                }

                // Add IPv6 support
                try {
                    builder.addAddress(TUN_IPV6_ADDR, TUN_IPV6_PREFIX)
                    builder.addRoute("::", 0)
                } catch (e: Exception) {
                    Log.w(TAG, "IPv6 route addition failed; falling back to IPv4-only", e)
                }

                // Kill Switch (System blocking mode)
                if (routingRepo.isKillSwitchEnabled && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    builder.setBlocking(true)
                }

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    builder.setMetered(false)
                }

                // Per-App Split Tunneling vs Global
                if (routingRepo.routingMode == RoutingMode.SPLIT_TUNNEL_PER_APP) {
                    val allowedApps = routingRepo.getSelectedApps()
                    if (allowedApps.isNotEmpty()) {
                        for (pkg in allowedApps) {
                            try {
                                builder.addAllowedApplication(pkg)
                            } catch (e: Exception) {
                                Log.w(TAG, "Could not add allowed app $pkg: ${e.message}")
                            }
                        }
                    } else {
                        // Fallback: exclude self
                        try {
                            builder.addDisallowedApplication(packageName)
                        } catch (_: Exception) {}
                    }
                } else {
                    // Global / Direct bypass: exclude self to prevent loop
                    try {
                        builder.addDisallowedApplication(packageName)
                    } catch (e: Exception) {
                        Log.w(TAG, "Failed to disallow self-package", e)
                    }
                }

                tunInterface = builder.establish()

                val fd = tunInterface?.fd
                if (fd != null && fd > 0) {
                    Log.i(TAG, "TUN interface established with FD: $fd")

                    // Step 5: Write YAML config for hev-socks5-tunnel C engine
                    val configFile = File(filesDir, "hev_tunnel_config.yaml")
                    val tunnelConfig = HevTunnelConfig.default(port = SOCKS5_PORT, mtu = TUN_MTU)
                    tunnelConfig.writeToFile(configFile)

                    // Step 6: Initialize and start native C tun2socks coroutine engine
                    val initResult = HevTunnel.init(configFile.absolutePath, fd)
                    if (initResult == 0) {
                        HevTunnel.start()
                        isRunning = true
                        Log.i(TAG, "AegisVPN pipeline fully running: TUN -> hev-socks5-tunnel -> Xray -> Remote Proxy")
                        withContext(Dispatchers.Main) {
                            startForeground(NOTIFICATION_ID, buildForegroundNotification("Connected • ${core.displayName} Active"))
                        }
                    } else {
                        Log.e(TAG, "HevTunnel nativeInit failed with code: $initResult")
                        stopVpn()
                    }
                } else {
                    Log.e(TAG, "Failed to acquire valid TUN file descriptor")
                    stopVpn()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error during VPN pipeline startup", e)
                stopVpn()
            }
        }
    }

    private fun stopVpn() {
        Log.i(TAG, "Stopping Aegis VPN Service and tearing down pipeline...")
        isRunning = false

        serviceScope.launch(Dispatchers.IO) {
            try {
                // 1. Stop hev-socks5-tunnel C loop
                HevTunnel.stop()

                // 2. Close Android TUN interface
                tunInterface?.close()
                tunInterface = null

                // 3. Stop loopback health monitor
                healthMonitor.stopMonitoring()

                // 4. Gracefully terminate protocol daemon (SIGTERM -> SIGKILL)
                processManager.stopDaemon()
            } catch (e: Exception) {
                Log.e(TAG, "Error closing tunnel pipeline", e)
            } finally {
                withContext(Dispatchers.Main) {
                    stopForeground(STOP_FOREGROUND_REMOVE)
                    stopSelf()
                }
            }
        }
    }

    private fun getOrCreateDefaultDaemonConfig(core: CoreBinary): File {
        val file = File(filesDir, "config_${core.executableName}.json")
        if (!file.exists()) {
            file.writeText(
                """
                {
                  "log": { "loglevel": "warning" },
                  "inbounds": [{
                    "port": $SOCKS5_PORT,
                    "listen": "127.0.0.1",
                    "protocol": "socks",
                    "settings": { "auth": "noauth", "udp": true }
                  }],
                  "outbounds": [{ "protocol": "freedom" }]
                }
                """.trimIndent()
            )
        }
        return file
    }

    private fun buildForegroundNotification(statusText: String): android.app.Notification {
        return NotificationCompat.Builder(this, AegisApplication.VPN_NOTIFICATION_CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_lock_lock)
            .setContentTitle(getString(R.string.app_name))
            .setContentText(statusText)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()
    }

    override fun onDestroy() {
        stopVpn()
        healthMonitor.release()
        processManager.release()
        serviceScope.cancel()
        super.onDestroy()
    }
}
